R/calculate_within_batch_ari.R

```diff
@@ -48,54 +48,96 @@ calculate_within_batch_ari <- function(individual_sce_list,
-        )
-
-      # Extract clusters from merged clustering for batch
-      clusters_to_keep <- grep(batch, merged_sce[[batch_column]])
-      batch_merged_clusters <- merged_clusters[clusters_to_keep]
-
-      # Calculate ARI between pre-merged clustering and post-merged clustering for the given batch
-      ari <- bluster::pairwiseRand(individual_clustering_result$individual_clusters,
-        batch_merged_clusters,
-        mode = "index"
+#' Calculate within-batch ARI using provided PCs for use in integration metric calculations
+#'
+#' @param merged_sce The merged SCE object containing data from multiple batches
+#' @param pc_name The name used to access the PCA results stored in the
+#'   SingleCellExperiment object
+#' @param batch_column The variable in `merged_sce` indicating the grouping of interest.
+#'  Generally this is either batches or cell types. Default is "library_id".
+#'
+#' @return Tibble with three columns: `ari`, representing the calculated ARI values;
+#'   `batch_id`, the batch id associated with each `ari`; `pc_name`, the name
+#'   associated with the pc results
+within_batch_ari_from_pcs <-
+  function(merged_sce, pc_name, batch_column = "library_id") {
+    # Check that `pc_name` is in merged SCE object
+    if (!pc_name %in% reducedDimNames(merged_sce)) {
+      stop("The provided `pc_name` cannot be found in the `merged_sce` object.")
+    }
+
+    # Check that `batch_column` is in colData of SCE
+    if (!batch_column %in% colnames(colData(merged_sce))) {
+      stop("The specified batch column is missing from the colData of the `merged_sce`.")
+    }
+
+    # Pull out the PCs or analogous reduction from merged object
+    merged_pcs <- reducedDim(merged_sce, pc_name)
+
+    # Cluster merged pcs only one time
+    merged_sce <- merged_sce |>
+      cluster_sce(
+        pc_name = pc_name,
+        BLUSPARAM = bluster::NNGraphParam(cluster.fun = "louvain", type = "jaccard"),
+        cluster_column_name = "merged_clusters"
       )
 
-      return(ari)
-    })
+    merged_clusters <- merged_sce$merged_clusters |>
+      purrr::set_names(merged_sce[[batch_column]])
 
-  # Create tibble with ARI and batch id
-  within_batch_ari_tibble <- tibble::tibble(
-    ari = all_ari,
-    batch_id = batch_ids,
-    pc_name = pc_name
-  )
+    # For every batch id, cluster and then calculate ARI for that batch
+    all_ari <- batch_ids |>
+      purrr::map_dbl(\(batch) {
+        # Cluster pc matrix for specified batch
+        individual_clustering_result <-
+          individual_sce_list[[batch]] |>
+          cluster_sce(
+            pc_name = "PCA",
+            BLUSPARAM = bluster::NNGraphParam(cluster.fun = "louvain", type = "jaccard"),
+            cluster_column_name = "individual_clusters"
+          )
 
-  return(within_batch_ari_tibble)
-}
+        # Extract clusters from merged clustering for batch
+        clusters_to_keep <- grep(batch, merged_sce[[batch_column]])
```

Comment:
I don't think you need grep here? And you are keeping cells, right, not whole clusters?
```suggestion
        cells_to_keep <- which(merged_sce[[batch_column]] == batch)
```

Reply:
Looking at this, I think I am actually a bit concerned that there is no guarantee that the order of the cells in these two SCE objects is the same. While we would not expect merging to alter the order, we don't guarantee that. 
Can we check that the barcodes all match between the individual clustering result and the batch clusters?

On line 106 you pull out the merged clusters, but I would move that down here to closer to where it will be used. But also there you set the names of the object to the batch names, which I'm not sure we need anymore? If not, maybe set it to the barcodes so we can check order?